nexus/test-utils/src/resource_helpers.rs

```diff
@@ -741,70 +751,117 @@ impl<'a, N: NexusServer> DiskTestBuilder<'a, N> {
-    pub async fn empty<N: NexusServer>(
-        cptestctx: &ControlPlaneTestContext<N>,
-    ) -> Self {
-        let sled_agent = cptestctx.sled_agent.sled_agent.clone();
+impl<'a> Iterator for ZpoolIterator<'a> {
+    type Item = &'a TestZpool;
 
-        Self { sled_agent, zpools: vec![] }
-    }
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let Some(sled_id) = self.sled else {
+                return None;
+            };
 
-    pub async fn new<N: NexusServer>(
-        cptestctx: &ControlPlaneTestContext<N>,
-    ) -> Self {
-        let sled_agent = cptestctx.sled_agent.sled_agent.clone();
+            let Some(pools) = self.sleds.get(&sled_id) else {
+                return None;
+            };
 
-        let mut disk_test = Self { sled_agent, zpools: vec![] };
+            if let Some(pool) = pools.zpools.get(self.index) {
+                self.index += 1;
+                return Some(pool);
+            }
 
-        // Create three Zpools, each 10 GiB, each with one Crucible dataset.
-        for _ in 0..Self::DEFAULT_ZPOOL_COUNT {
-            disk_test.add_zpool_with_dataset(cptestctx).await;
+            self.sled = self
+                .sleds
+                .range((
+                    std::ops::Bound::Excluded(&sled_id),
+                    std::ops::Bound::Unbounded,
+                ))
+                .map(|(id, _)| *id)
+                .next();
+            self.index = 0;
         }
+    }
+}
 
-        disk_test
+pub struct DiskTest<'a, N: NexusServer> {
+    cptestctx: &'a ControlPlaneTestContext<N>,
+    sleds: BTreeMap<SledUuid, PerSledDiskState>,
+}
+
+impl<'a, N: NexusServer> DiskTest<'a, N> {
+    pub const DEFAULT_ZPOOL_SIZE_GIB: u32 = 10;
+    pub const DEFAULT_ZPOOL_COUNT: u32 = 3;
+
+    /// Creates a new [DiskTest] with default configuration.
+    ///
+    /// This is the same as calling [DiskTestBuilder::new] and
+    /// [DiskTestBuilder::build].
+    pub async fn new(cptestctx: &'a ControlPlaneTestContext<N>) -> Self {
+        DiskTestBuilder::new(cptestctx).build().await
     }
 
-    pub async fn new_from_builder<N: NexusServer>(
-        cptestctx: &ControlPlaneTestContext<N>,
-        sled_agent: Arc<SledAgent>,
+    async fn new_from_builder(
+        cptestctx: &'a ControlPlaneTestContext<N>,
+        sled_agents: WhichSledAgents,
         zpool_count: u32,
     ) -> Self {
-        let mut disk_test = Self { sled_agent, zpools: vec![] };
+        let input_sleds = match sled_agents {
+            WhichSledAgents::Specific(id) => {
+                vec![id]
+            }
+            WhichSledAgents::All => {
+                vec![
+                    SledUuid::from_untyped_uuid(
+                        cptestctx.sled_agent.sled_agent.id,
+                    ),
+                    SledUuid::from_untyped_uuid(
+                        cptestctx.sled_agent2.sled_agent.id,
+                    ),
+                ]
```

Comment:
It looks like there's a `cptestctx.all_sled_agents()`, you could `map` over that to get the sled agent ids?

Reply:
ah, yup, good call. I added `all_sled_agents` in this PR to avoid hard-coded "two sled" enumerations like this.